manifest/list.go

```diff
@@ -0,0 +1,118 @@
+package manifest
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/containers/image/types"
+	digest "github.com/opencontainers/go-digest"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// List is an interface for parsing, modifying lists of image manifests.
+// Callers can either use this abstract interface without understanding the
+// details of the formats, or instantiate a specific implementation (e.g.
+// manifest.OCI1Index) and access the public members directly.
+type List interface {
+	// MIMEType returns the MIME type of this particular manifest list.
+	MIMEType() string
+
+	// Instances returns a list of the manifests that this list knows of.
+	Instances() []types.BlobInfo
+
+	// Update information about the list's instances.  The length of the passed-in slice must
+	// match the length of the list of instances which the list already contains, and every field
+	// must be specified.
+	UpdateInstances([]ListUpdate) error
+
+	// ChooseDigest selects which manifest is most appropriate for the platform described by the
+	// SystemContext, or for the current platform if the SystemContext doesn't specify any details.
+	ChooseDigest(ctx *types.SystemContext) (digest.Digest, error)
```

Comment:
This PR has changed `ChooseInstance` to `ChooseDigest` here and in a few other places, but e.g. the `Instances` and `UpdateInstances` names still exist; all of that should be consistent.

(I weakly prefer `Instance` because hard-coding the concept of sha256 digests throughout the codebase seems shakier as formats evolve. OTOH that could be an argument for using the `Digest` name, and reserving `Instance` for some future more general abstraction, if it turned out to be necessary.)

https://github.com/containers/image/pull/400#discussion_r163015366